rust/agents/validator/src/submit.rs

```diff
@@ -58,33 +58,92 @@ impl ValidatorSubmitter {
+        }
+    }
+
+    /// Submits signed checkpoints indefinitely, starting from the `tree`.
+    #[instrument(err, skip(self, tree), fields(domain=%self.merkle_tree_hook.domain()))]
+    pub(crate) async fn checkpoint_submitter(self, mut tree: IncrementalMerkle) -> Result<()> {
+        loop {
+            // Lag by reorg period because this is our correctness checkpoint.
+            let latest_checkpoint = self
+                .merkle_tree_hook
+                .latest_checkpoint(self.reorg_period)
+                .await?;
+            self.metrics
+                .latest_checkpoint_observed
+                .set(latest_checkpoint.index as i64);
+
+            // This may occur e.g. if RPC providers are unreliable and make calls against
+            // inconsistent block tips.
+            //
+            // In this case, we just sleep a bit until we fetch a new latest checkpoint
+            // that at least meets the tree.
+            if tree_exceeds_checkpoint(&latest_checkpoint, &tree) {
+                debug!(
+                    ?latest_checkpoint,
+                    tree_count = tree.count(),
+                    "Latest checkpoint is behind tree, sleeping briefly"
+                );
+                sleep(self.interval).await;
+                continue;
+            }
 
-            // ingest available messages from DB
-            while let Some(insertion) = self
+            self.submit_checkpoints_until_correctness_checkpoint(&mut tree, &latest_checkpoint)
+                .await?;
+
+            self.metrics
+                .latest_checkpoint_processed
+                .set(latest_checkpoint.index as i64);
+
+            sleep(self.interval).await;
+        }
+    }
+
+    /// Submits signed checkpoints relating to the given tree until the correctness checkpoint (inclusive).
+    /// Only submits the signed checkpoints once the correctness checkpoint is reached.
+    async fn submit_checkpoints_until_correctness_checkpoint(
+        &self,
+        tree: &mut IncrementalMerkle,
+        correctness_checkpoint: &Checkpoint,
+    ) -> Result<()> {
+        // This should never be called with a tree that is ahead of the correctness checkpoint.
+        assert!(
+            !tree_exceeds_checkpoint(correctness_checkpoint, tree),
+            "tree (count: {}) is ahead of correctness checkpoint {:?}",
+            tree.count(),
+            correctness_checkpoint,
+        );
+
+        // All intermediate checkpoints will be stored here and signed once the correctness
+        // checkpoint is reached.
+        let mut checkpoint_queue = vec![];
+
+        // If the correctness checkpoint is ahead of the tree, we need to ingest more messages.
+        //
+        // tree.index() will panic if the tree is empty, so we use tree.count() instead
+        // and convert the correctness_checkpoint.index to a count by adding 1.
+        while correctness_checkpoint.index + 1 > tree.count() as u32 {
```

Comment:
is there no `tree.index()`?

Reply:
can you use `tree_exceeds_checkpoint` here?

Reply:
See the comment above on tree.index(), it panics if the count is 0 so this is the unfortunate workaround

And I don't think so because `tree_exceeds_checkpoint` returns true if tree.index > checkpoint.index (so doing `!tree_exceeds_checkpoint` would be if tree.index <= checkpoint.index) and we want the condition here to be checkpoint.index > tree.index